# Project 5: Deadpan Dog Analysis
An analysis, *sans* emotion, of the overemotional rating of dogs by the WeRateDogs twitter account.

*This project is for Unit 4, Data Wrangling, of the Data Analyst Nanodegree from Udacity*

## Table of Contents
<ul>
<li><a href='#intro'>Introduction</a></li>
<li><a href='#wrangling'>Data Wrangling</a></li>
    <ul>
    <li><a href='#gather'>Gathering Data</a></li>
    <li><a href='#assess'>Assessing Data</a></li>
    <li><a href='#clean'>Cleaning Data</a></li>
    </ul>
<li><a href='#eda'>Exploratory Data Analysis</a></li>
<li><a href='#stats'>Statistical Analysis</a></li>
<li><a href='#conclusions'>Conclusions</a></li>
</ul>

<a id='intro'></a>
## Introduction
In this project, I will be working with the WeRateDogs Twitter account's tweets. This is a combination of the following sources:

- Archived Twitter Data (provided as a "file on hand" by Udacity)
- Image Predictions for Tweets (requested from a Udacity-hosted neural network)
- Additional Tweet Metadata (queried via Twitter API)

### Questions
1. What are Dogs?
2. How to doggo?
3. Why floof?

In [348]:
#importing necessary modules
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import requests
import tweepy
import json
from timeit import default_timer as timer

<a id='wrangling'></a>
## Data Wrangling
<ul>
<li><a href='#gather'>Gathering Data</a></li>
<li><a href='#assess'>Assessing Data</a></li>
<li><a href='#clean'>Cleaning Data</a></li>
</ul>

<a id='gather'></a>
### Gathering Data

In [349]:
#reading in the Archived Twitter Data file into a data frame
df_archive = pd.read_csv('twitter-archive-enhanced.csv')

In [350]:
#pulling Image Predictions from Udacity-hosted file online
# url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
# r = requests.get(url)
# with open(url.split('/')[-1], 'wb') as file:
#     file.write(r.content)

In [351]:
#reading in the newly created Image Predictions file as a data frame
df_predict = pd.read_csv('image-predictions.tsv', sep='\t')

In [352]:
#scraping metadata JSON using twitter API
#Note: the four variables used in the arguments for lines 4 and 5 were removed for privacy
# start = timer()
# auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
# auth.set_access_token(access_token, access_token_secret)

# api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

# lost_tweets = []
# with open('tweet_json.txt', 'w') as file:
#     for tweet_id in df_archive.tweet_id:
#         try:
#             tweet = api.get_status(tweet_id, tweet_mode='extended')
#             json.dump(tweet._json, file)
#             file.write('\n')
#         except tweepy.TweepError as te:
#                 lost_tweets.append(tweet_id)
# end = timer()
# print('Time Elapsed:', end - start)

In [353]:
#reading in the newly created JSON file of tweet metadata as a data frame
df_meta = pd.read_json('tweet_json.txt', lines=True)

<a id='assess'></a>
### Assessing Data

In [354]:
df_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [355]:
df_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [356]:
df_archive.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [357]:
df_archive[df_archive.rating_denominator != 10].head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313  835246439529840640           8.352460e+17           26259576.0   
342  832088576586297345           8.320875e+17           30582082.0   
433  820690176645140481                    NaN                  NaN   
516  810984652412424192                    NaN                  NaN   
784  775096608509886464                    NaN                  NaN   

                     timestamp  \
313  2017-02-24 21:54:03 +0000   
342  2017-02-16 04:45:50 +0000   
433  2017-01-15 17:52:40 +0000   
516  2016-12-19 23:06:23 +0000   
784  2016-09-11 22:20:06 +0000   

                                                source  \
313  <a href="http://twitter.com/download/iphone" r...   
342  <a href="http://twitter.com/download/iphone" r...   
433  <a href="http://twitter.com/download/iphone" r...   
516  <a href="http://twitter.com/download/iphone" r...   
784  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
313  @jonnysun @Lin_Manuel ok jomny I know you're e...                  NaN   
342           @docmisterio account started on 11/15/15                  NaN   
433  The floofs have been released I repeat the flo...                  NaN   
516  Meet Sam. She smiles 24/7 &amp; secretly aspir...                  NaN   
784  RT @dog_rates: After so many requests, this is...         7.403732e+17   

     retweeted_status_user_id retweeted_status_timestamp  \
313                       NaN                        NaN   
342                       NaN                        NaN   
433                       NaN                        NaN   
516                       NaN                        NaN   
784              4.196984e+09  2016-06-08 02:41:38 +0000   

                                         expanded_urls  rating_numerator  \
313                                                NaN               960   
342                                                NaN                11   
433  https://twitter.com/dog_rates/status/820690176...                84   
516  https://www.gofundme.com/sams-smile,https://tw...                24   
784  https://twitter.com/dog_rates/status/740373189...                 9   

     rating_denominator  name doggo floofer pupper puppo  
313                   0  None  None    None   None  None  
342                  15  None  None    None   None  None  
433                  70  None  None    None   None  None  
516                   7   Sam  None    None   None  None  
784                  11  None  None    None   None  None

In [358]:
#investigating reason for wonky rating_numerator or rating_denominator
print(df_archive.iloc[313].text)

@jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho


In [359]:
#investigating source looking so strange
df_archive.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

In [360]:
#seeing if dog "stage" columns are mutually exclusive
df_archive[['doggo', 'floofer', 'pupper', 'puppo']].query('pupper != "None" & doggo != "None"')

doggo floofer  pupper puppo
460   doggo    None  pupper  None
531   doggo    None  pupper  None
565   doggo    None  pupper  None
575   doggo    None  pupper  None
705   doggo    None  pupper  None
733   doggo    None  pupper  None
778   doggo    None  pupper  None
822   doggo    None  pupper  None
889   doggo    None  pupper  None
956   doggo    None  pupper  None
1063  doggo    None  pupper  None
1113  doggo    None  pupper  None

In [361]:
df_archive.iloc[889].text

'Meet Maggie &amp; Lila. Maggie is the doggo, Lila is the pupper. They are sisters. Both 12/10 would pet at the same time https://t.co/MYwR4DQKll'

In [362]:
df_predict.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [363]:
df_predict.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [364]:
df_meta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 32 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   created_at                     2331 non-null   datetime64[ns, UTC]
 1   id                             2331 non-null   int64              
 2   id_str                         2331 non-null   int64              
 3   full_text                      2331 non-null   object             
 4   truncated                      2331 non-null   bool               
 5   display_text_range             2331 non-null   object             
 6   entities                       2331 non-null   object             
 7   extended_entities              2059 non-null   object             
 8   source                         2331 non-null   object             
 9   in_reply_to_status_id          77 non-null     float64            
 10  in_reply_to_status_id_st

In [365]:
df_meta.head()

created_at                  id              id_str  \
0 2017-08-01 16:23:56+00:00  892420643555336193  892420643555336192   
1 2017-08-01 00:17:27+00:00  892177421306343426  892177421306343424   
2 2017-07-31 00:18:03+00:00  891815181378084864  891815181378084864   
3 2017-07-30 15:58:51+00:00  891689557279858688  891689557279858688   
4 2017-07-29 16:00:24+00:00  891327558926688256  891327558926688256   

                                           full_text  truncated  \
0  This is Phineas. He's a mystical boy. Only eve...      False   
1  This is Tilly. She's just checking pup on you....      False   
2  This is Archie. He is a rare Norwegian Pouncin...      False   
3  This is Darla. She commenced a snooze mid meal...      False   
4  This is Franklin. He would like you to stop ca...      False   

  display_text_range                                           entities  \
0            [0, 85]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1           [0, 138]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2           [0, 121]  {'hashtags': [], 'symbols': [], 'user_mentions...   
3            [0, 79]  {'hashtags': [], 'symbols': [], 'user_mentions...   
4           [0, 138]  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   

                                   extended_entities  \
0  {'media': [{'id': 892420639486877696, 'id_str'...   
1  {'media': [{'id': 892177413194625024, 'id_str'...   
2  {'media': [{'id': 891815175371796480, 'id_str'...   
3  {'media': [{'id': 891689552724799489, 'id_str'...   
4  {'media': [{'id': 891327551943041024, 'id_str'...   

                                              source  in_reply_to_status_id  \
0  <a href="http://twitter.com/download/iphone" r...                    NaN   
1  <a href="http://twitter.com/download/iphone" r...                    NaN   
2  <a href="http://twitter.com/download/iphone" r...                    NaN   
3  <a href="http://twitter.com/download/iphone" r...                    NaN   
4  <a href="http://twitter.com/download/iphone" r...                    NaN   

   ...  favorited  retweeted  possibly_sensitive  \
0  ...      False      False                 0.0   
1  ...      False      False                 0.0   
2  ...      False      False                 0.0   
3  ...      False      False                 0.0   
4  ...      False      False                 0.0   

  possibly_sensitive_appealable lang  retweeted_status  quoted_status_id  \
0                           0.0   en               NaN               NaN   
1                           0.0   en               NaN               NaN   
2                           0.0   en               NaN               NaN   
3                           0.0   en               NaN               NaN   
4                           0.0   en               NaN               NaN   

  quoted_status_id_str  quoted_status_permalink  quoted_status  
0                  NaN                      NaN            NaN  
1                  NaN                      NaN            NaN  
2                  NaN                      NaN            NaN  
3                  NaN                      NaN            NaN  
4                  NaN                      NaN            NaN  

[5 rows x 32 columns]

In [366]:
#investigating id vs id_str
df_predict[df_predict.tweet_id == 892420643555336193]

tweet_id                                          jpg_url  \
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num      p1   p1_conf  p1_dog     p2   p2_conf  p2_dog      p3  \
2074        1  orange  0.097049   False  bagel  0.085851   False  banana   

      p3_conf  p3_dog  
2074  0.07611   False

#### Issues

##### Quality

1. Tweet id is referred to inconsistently across tables
2. There are many empty columns and columns that aren't needed (and are therefore invalid)

`df_archive`
3. Archived data contains 181 retweets, which are invalid based on our desire for original tweets only
4. `rating_denominator` data should all be 10, but some are very far off, most due to scraping the wrong numbers from the text.
5. `rating_numerator` values have a similar problem to the denominator
6. `timestamp` is object type, and should be datetime type.
7. `source` is invalid as it contains the whole hyperlink html tag. This should be a simpler categorical data format preferably a one word string.

`df_meta`
8. Data is incomplete since some tweets were not able to be pulled by the API (IDs were stored in `lost_tweets` and can also be found comparing to df_archive)
9. `id` does not match `id_str` for some tweets, and `id` matches the id used in the other two tables.

##### Tidiness
1. Data is split across three dataframes and three files, but only has two main observational units, i.e., tweet data and tweet image predictions
2. Duplicated information (tweet id in 4 columns, text in both `df_archive` and `df_meta`)
3. In `df_archive`, the dog "stage" descriptions should be one column instead of 4 since it's mostly exlusive categorical data

<a id='clean'></a>
### Cleaning Data

In [367]:
#making clean copies to work with. `df_tweet` will be the clean combined version of df_archive and df_meta
df_tweet = df_archive.copy()
df_meta_clean = df_meta.copy()
df_predict_clean = df_predict.copy()

In [368]:
df_tweet.set_index('tweet_id', inplace=True)

#### `df_archive` Archived data contains 181 retweets, which are invalid based on our desire for original tweets only. Data is incomplete since some tweets were not able to be pulled by the API (IDs were stored in lost_tweets and can also be found comparing to df_archive).

##### Describe
Drop rows from `df_archive` using the `retweeted_status_id` column as an indicator for being a retweet.

##### Code

In [369]:
df_tweet.drop(df_tweet[df_tweet.retweeted_status_id.notna()].index, inplace=True)

##### Test

In [370]:
df_tweet.shape

(2175, 16)

In [371]:
df_tweet.retweeted_status_id.isna().sum()

2175

#### Tweet id is referred to inconsistently across tables
##### Describe
Rename the `id` column in `df_meta` to be `tweet_id` to match the other two tables.
##### Code

In [372]:
df_meta_clean.rename(columns={'id':'tweet_id'}, inplace=True)

In [373]:
df_meta_clean.head(1)

created_at            tweet_id              id_str  \
0 2017-08-01 16:23:56+00:00  892420643555336193  892420643555336192   

                                           full_text  truncated  \
0  This is Phineas. He's a mystical boy. Only eve...      False   

  display_text_range                                           entities  \
0            [0, 85]  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                   extended_entities  \
0  {'media': [{'id': 892420639486877696, 'id_str'...   

                                              source  in_reply_to_status_id  \
0  <a href="http://twitter.com/download/iphone" r...                    NaN   

   ...  favorited  retweeted  possibly_sensitive  \
0  ...      False      False                 0.0   

  possibly_sensitive_appealable lang  retweeted_status  quoted_status_id  \
0                           0.0   en               NaN               NaN   

  quoted_status_id_str  quoted_status_permalink  quoted_status  
0                  NaN                      NaN            NaN  

[1 rows x 32 columns]

#### There are many empty columns and columns that aren't needed (and are therefore invalid), in all three df
##### Describe
Redefine `df_meta_clean` and drop columns in `df_tweet` to only include the columns that will be useful in analysis later
##### Code

In [374]:
df_meta_clean = df_meta_clean.copy()[['tweet_id', 'created_at', 'retweet_count', 'favorite_count']]
df_tweet.drop(columns=['in_reply_to_status_id','in_reply_to_user_id','retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp'],inplace=True)

##### Test

In [375]:
df_meta_clean.head(1)

tweet_id                created_at  retweet_count  favorite_count
0  892420643555336193 2017-08-01 16:23:56+00:00           7650           35960

In [376]:
df_tweet.head(1)

timestamp  \
tweet_id                                        
892420643555336193  2017-08-01 16:23:56 +0000   

                                                               source  \
tweet_id                                                                
892420643555336193  <a href="http://twitter.com/download/iphone" r...   

                                                                 text  \
tweet_id                                                                
892420643555336193  This is Phineas. He's a mystical boy. Only eve...   

                                                        expanded_urls  \
tweet_id                                                                
892420643555336193  https://twitter.com/dog_rates/status/892420643...   

                    rating_numerator  rating_denominator     name doggo  \
tweet_id                                                                  
892420643555336193                13                  10  Phineas  None   

                   floofer pupper puppo  
tweet_id                                 
892420643555336193    None   None  None

#### Data is split across three dataframes and three files, but only has two main observational units, i.e., tweet data and tweet image predictions

##### Describe
Merge in the timestamp, retweet count, and favorite count data from `df_meta`. This will also incidentally fix the id vs id_str problem and eliminate having to deal with the extra columns in `df_meta`, and the tidiness issue of duplicate information (like id in 4 places). Merging "inner" also avoids the problem of deleted tweets in the archived data by only keeping the tweets that were able to be pulled by the API.
##### Code

In [377]:
df_tweet = pd.merge(df_tweet, df_meta_clean, on='tweet_id')

##### Test

In [378]:
df_tweet.shape

(2168, 15)

In [379]:
df_tweet.head(1)

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  \
0                  10  Phineas  None    None   None  None   

                 created_at  retweet_count  favorite_count  
0 2017-08-01 16:23:56+00:00           7650           35960

#### `timestamp` is object type, and should be datetime type.
##### Describe
Drop `timestamp` column and use `created_at` instead
##### Code

In [380]:
df_tweet.drop(columns='timestamp', inplace=True)

##### Test

In [381]:
df_tweet.columns

Index(['tweet_id', 'source', 'text', 'expanded_urls', 'rating_numerator',
       'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo',
       'created_at', 'retweet_count', 'favorite_count'],
      dtype='object')

#### In `df_archive`, The dog "stage" descriptions should be one column instead of 4. 
##### Describe
Replace None with empty strings and then make a new column for `dog_stage` concatenating the columns. Replace any non-mutually exclusive data with "multiple" and drop the existing 4 columns.
##### Code

In [382]:
df_tweet.replace(to_replace='None', value='', inplace=True)
df_tweet['dog_stage'] = (df_tweet.doggo + df_tweet.floofer + df_tweet.pupper + df_tweet.puppo)

In [383]:
df_tweet.dog_stage.replace(to_replace=['doggopupper','doggopuppo','doggofloofer'], value='multiple', inplace=True)
df_tweet.dog_stage.replace(to_replace='', value=np.nan, inplace=True)

In [384]:
df_tweet.drop(columns=['doggo','floofer','pupper','puppo'], inplace=True)

##### Test

In [385]:
df_tweet.dog_stage.value_counts()

pupper      224
doggo        74
puppo        24
multiple     12
floofer       9
Name: dog_stage, dtype: int64

In [386]:
df_tweet.head(1)

tweet_id                                             source  \
0  892420643555336193  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name                created_at  retweet_count  \
0                  10  Phineas 2017-08-01 16:23:56+00:00           7650   

   favorite_count dog_stage  
0           35960       NaN

#### `source` column is invalid as it contains the whole hyperlink html tag. This should be a simpler categorical data format preferably a one word string.
##### Describe
Use the pd.Series.str.replace() method to find the identifying informatoin about the source of the Tweet, and discard the rest.
##### Code

In [387]:
replacements = dict(zip(list(df_tweet.source.value_counts().index), ['iPhone','Vine','Website','Desktop']))
df_tweet.source.replace(replacements, inplace=True)

##### Test

In [388]:
df_tweet.source.value_counts()

iPhone     2036
Vine         91
Website      31
Desktop      10
Name: source, dtype: int64

#### `rating_denominator` data should all be 10, but some are very far off, most due to scraping the wrong numbers from the text. `rating_numerator` values have a similar problem to the denominator. Some may have not have ratings, and some may just be pulling the wrong number.
##### Describe
Since only 22 rows of the 2168 row dataset remaining have these types of issues, I'm just going to drop the affected rows (of denominators >10). The amount of effort and time required would not be worth it to either write a function to programmatically select the correct numbers for each situation or go through all 22 manually. I will check the numerators after dropping rows to see if there are large outliers there still.
##### Code

In [389]:
df_tweet.drop(df_tweet[df_tweet.rating_denominator != 10].index, inplace=True)

In [343]:
df_tweet[df_tweet.rating_numerator > 10].rating_numerator.value_counts()

12      498
11      425
13      303
14       43
420       2
1776      1
666       1
182       1
75        1
27        1
26        1
17        1
15        1
Name: rating_numerator, dtype: int64

These remaining outliers are still too much. 1776 will really skew the data, although it is technically a valid rating. I think the analysis will be more poignant and informative without those outliers, although they're interesting and I will note them in my report.

In [344]:
df_tweet.drop(df_tweet.query('rating_numerator > 50').index, inplace=True)

##### Test

In [345]:
df_tweet[df_tweet.rating_denominator != 10].tweet_id.count()

0

In [390]:
df_tweet[df_tweet.rating_numerator > 50].text.values

array(['@dhmontgomery We also gave snoop dogg a 420/10 but I think that predated your research',
       '@s8n You tried very hard to portray this good boy as not so good, but you have ultimately failed. His goodness shines through. 666/10',
       '@markhoppus 182/10',
       "This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wuqaPS",
       "This is Atticus. He's quite simply America af. 1776/10 https://t.co/GRXwMxLBkh",
       'After so many requests... here you go.\n\nGood dogg. 420/10 https://t.co/yfAAo1gdeY'],
      dtype=object)

In [347]:
df_tweet.shape

(2140, 11)

<a id='eda'></a>
## Exploratory Data Analysis

<a id='stats'></a>
## Statistical Analysis

<a id='conclusions'></a>
## Conclusions